core/src/main/kotlin/com/playkids/onboarding/core/service/ProfileService.kt

```diff
@@ -0,0 +1,64 @@
+package com.playkids.onboarding.core.service
+
+import com.playkids.onboarding.core.excption.EntityNotFoundException
+import com.playkids.onboarding.core.excption.NotEnoughCurrency
+import com.playkids.onboarding.core.excption.UserHasItemException
+import com.playkids.onboarding.core.model.ItemId
+import com.playkids.onboarding.core.model.Profile
+import com.playkids.onboarding.core.model.ProfileId
+import com.playkids.onboarding.core.model.SKUId
+import com.playkids.onboarding.core.persistence.ItemDAO
+import com.playkids.onboarding.core.persistence.ProfileDAO
+import com.playkids.onboarding.core.persistence.SKUDAO
+import com.playkids.onboarding.core.util.ChooseValue
+
+class ProfileService(
+    private val profileDAO: ProfileDAO,
+    private val skuDAO: SKUDAO,
+    private val itemDAO: ItemDAO
+) {
+    suspend fun create(profile: Profile){
+        profileDAO.create(profile)
+    }
+
+    suspend fun find(id: ProfileId): Profile?{
+        return profileDAO.find(id)
+    }
+
+    suspend fun addItem(profileId: ProfileId, itemId: ItemId, itemCategory: String){
+        return profileDAO.addItem(profileId, listOf("$itemCategory:$itemId"))
+    }
+
+    suspend fun buyItem(profileId: ProfileId, itemId: ItemId, itemCategory: String, currency: String){
+        val item = itemDAO.find(itemCategory, itemId) ?: throw EntityNotFoundException("Item with id $itemId and category $itemCategory doesn't exists")
+        val price = when(currency){
+            COIN -> item.coinPrice
+            GEM -> item.gemPrice
+            else -> {
+                throw IllegalArgumentException("currency $currency doesn't exists")
+            }
+        } ?: throw IllegalArgumentException("Item doesn't have attribute $currency")
```

Comment:
Currency could be an enum treated elsewhere. You could also extract this part of the code to a separate method

Reply:
I don't think that the currency should be part of the request, you are delegating to the client the responsibility to decide whether the coin or the gem price will be used (which can lead to an exploit). If the item always gonna have one of the two prices (not a mixed one), the entity could have a currency enum and a price field.

Reply:
ahh I thought that an item could have a coin and a gem price! I will change it so it makes more sense in the case where it can only have one

Reply:
I'm not saying that the items can't have a mixed price ahahah, but the way you're doing it today forces it to have only one, as you have to choose a currency in this switch/case. In this case, I think that makes sense to assume that the price is going to be only in gems or in coins.

Reply:
hahah fair enough! I changed it so each item can only be bought using coins or gems. It made the code way prettier and more readable